stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -0,0 +1,109 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tracing implements the OpenTelemetry carrier for context propagation
+// in gRPC tracing.
+package tracing
+
+import (
+	"context"
+
+	otelpropagation "go.opentelemetry.io/otel/propagation"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+// GRPCTraceBinHeaderKey is the gRPC metadata header key `grpc-trace-bin` used
+// to propagate trace context in binary format.
+const GRPCTraceBinHeaderKey = "grpc-trace-bin"
+
+// CustomCarrier is a TextMapCarrier that uses gRPC context to store and
+// retrieve any propagated key-value pairs in text format along with binary
+// format for `grpc-trace-bin` header
+type CustomCarrier struct {
+	otelpropagation.TextMapCarrier
+
+	ctx context.Context
+}
+
+// NewCustomCarrier creates a new CustomMapCarrier with
+// the given context.
+func NewCustomCarrier(ctx context.Context) *CustomCarrier {
+	return &CustomCarrier{
+		ctx: ctx,
+	}
+}
+
+// Get returns the string value associated with the passed key from the gRPC
+// context. It returns an empty string if the key is not present in the
+// context.
+func (c *CustomCarrier) Get(key string) string {
+	md, ok := metadata.FromIncomingContext(c.ctx)
+	if !ok {
+		return ""
+	}
+	values := md.Get(key)
+	if len(values) == 0 {
+		return ""
+	}
+	return values[0]
+}
+
+// Set stores the key-value pair in string format in the gRPC context.
+// If the key already exists, its value will be overwritten.
+func (c *CustomCarrier) Set(key, value string) {
+	md, ok := metadata.FromOutgoingContext(c.ctx)
+	if !ok {
+		md = metadata.MD{}
+	}
+	md.Set(key, value)
+	c.ctx = metadata.NewOutgoingContext(c.ctx, md)
```

Comment:
If these sets and gets get provided a "grpc-trace-bin", this needs to base 64 encode/decode. You probably should also log if you get a header that ends with bin that isn't "grpc-trace-bin".

Reply:
Actually, this CustomCarrier is supported for any propagator and not just GRPCTraceBinPropagator. As per design this carrier has to support both grpc-trace-bin and any other text based carrier so specifically for Get/Set which are interface methods should support any text key-value pair. It doesn't care about encoding/key names.

Reply:
GRPCTraceBinPropagator, on the other hand, is a propagator implementation GRPC provides to propagate `grpc-trace-bin` which can be called with either CustomCarrier or any other Carrier. If CustomCarrier is used with GRPCTraceBinPropagator that means its invoked from internal gRPC code so it will always use the binary path. For code outside of gRPC library, user can use the GRPCTraceBinPropagator with any other text based carrier.

Reply:
Java and C had this as part of their implementation. This is needed to keep it working smoothly.

Reply:
I looked at the Java implementation and they have special handling for grpc-trace-bin in carrier for regular get and set. Basically, we want to ensure if any other propagator calls CustomCarrier to get/set `grpc-trace-bin`, we should handle that the similar way with addition of base64 encoding/decoding.

I have modified the code to handle the above case and added tests for the same. Thanks for calling it out